### PR TITLE
Add copy link after token been sent for the user doesnt have email set up from design

### LINKED ISSUE #19625: Show user invite link when inviting user
At the moment the only way to invite a user is by letting Matomo send an E-Mail to them. But a ton of people have broken E-Mail setups. Either their mail server isn't properly set up with DKIM, etc. and all mails are rejected as SPAM or they don't really have a mail server or they are using some wonky SMTP setup via gmail (which can break for all kinds of Google changes).
Before that didn't really matter as unless you need E-Mail reports, you are only missing out on some things (mostly password reset and security notifications). 
But now with the invite feature Matomo becomes completely unusable unless your Matomo instance is able to successfully and reliably send E-Mails. And I am sure that will trip up and frustrate a lot of newer Matomo uses.

One simple solution would be to display the user invite link that is sent to the new user also to the admin when doing the invite. This also allows the admin to also send the link via other ways than E-Mail and allows people to easily circumvent the invite feature (assuming it works when already signed in).

### PR DESCRIPTION
**Pull request by peterhashair**
### Description:

Fixes: #19625

Add copy link after the token has been sent for the user who doesn't have an email set up from design. 

- As per the design, add a popup to resend invites. 
- As suggested, add an additional database field for the link token, and make 2 option works. 
- Add a new API endpoint.
- Add 2 new Icon, arrow left and arrow right
- As suggested, when over HTTPS using js `WriteText` otherwise `executeCommand('copy')`, there is still a security risk in `WriteText`, when failed, show the whole link.

This problem needs to update docs as well.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/Morpheus/stylesheets/ui/_buttons.less
Comment (sgiehl) [inline plugins/Morpheus/stylesheets/ui/_buttons.less]: That really doesn't look like a style that should be located in a global class. An absolute positioned element isn't anything that can be reused.

File: plugins/Morpheus/stylesheets/ui/_buttons.less
Comment (sgiehl) [inline plugins/Morpheus/stylesheets/ui/_buttons.less]: Where do you have that color from? We should try to reuse the common colors defined in https://github.com/matomo-org/matomo/blob/4.x-dev/plugins/Morpheus/stylesheets/base/colors.less, as they can't be overwritten by themes otherwise.

File: plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue
Comment (sgiehl) [inline plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue]: please avoid using scoped styles.

File: core/Db/Schema/Mysql.php
Comment (sgiehl) [inline core/Db/Schema/Mysql.php:58]: Having two columns for an invite token, but only one for expiry doesn't make any sense at all. That means you extend the expiry of the token send by mail if you copy a new one in the UI.
I would still prefer the solution of only having one valid token. As long as it's explained in the UI, that generating or sending a new token invalidates the previous one, that should be fine.

**Review by justinvelluppillai — CHANGES_REQUESTED** [review]
I left a few screenshots and comments for you to get it conforming to spec in slack 👍🏽

File: plugins/UsersManager/API.php
Comment (sgiehl) [inline plugins/UsersManager/API.php]: Then I guess we need to discuss the design again. Why should this be removed again, we actually just added it 🤷

File: plugins/Morpheus/stylesheets/base/icons.css
Comment (sgiehl) [inline plugins/Morpheus/stylesheets/base/icons.css]: new icons, if needed, should always be added to our demo page. Here https://github.com/matomo-org/matomo/blob/4.x-dev/plugins/Morpheus/templates/demo.twig#L621

Also I wonder if we directly should add the same arrow in the right direction.

File: plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue
Comment (sgiehl) [inline plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue:526]: Any reason for removing the `Edit user` in the headline?

File: plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue
Comment (sgiehl) [inline plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue]: shouldn't that one use the config value, instead of a hard coded `7`?

File: plugins/Morpheus/stylesheets/ui/_buttons.less
Comment (sgiehl) [inline plugins/Morpheus/stylesheets/ui/_buttons.less]: This is a very global css class, but it contains very specific rules. The class is not reusable that way, or do you think every close button should be placed absolute? Might be better to move that where it is used and make the css selector more specific.

File: plugins/UsersManager/tests/System/expected/test___UsersManager.getUsersWithSiteAccess_3_admin_when_superuseraccess.xml
Comment (sgiehl) [inline plugins/UsersManager/tests/System/expected/test___UsersManager.getUsersWithSiteAccess_3_admin_when_superuseraccess.xml]: The invite token are for sure nothing that should be returned in any API response. 
The are even stored hashed in database, so there is no use for having that at all. Those should be hiddne like the password and other fields. e.g. here https://github.com/matomo-org/matomo/blob/45058209303f51216367f07fde54f83a6928e1e6/plugins/UsersManager/Repository/UserRepository.php#L143-L153

File: plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue
Comment (sgiehl) [inline plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue]: Please ALWAYS avoid commenting out code that shouldn't be used anymore. It does not have any benefit keeping such stuff. If it needs to be reverted at some point, that can be easily done using git history. No need for polluting code.

File: plugins/UsersManager/lang/en.json
Comment (sgiehl) [inline plugins/UsersManager/lang/en.json]:
```suggestion
        "SendInvite": "Send invite",
```

File: plugins/UsersManager/lang/en.json
Comment (sgiehl) [inline plugins/UsersManager/lang/en.json]:
```suggestion
        "CopyLink": "Copy link",
```

File: plugins/UsersManager/lang/en.json
Comment (sgiehl) [inline plugins/UsersManager/lang/en.json]:
```suggestion
        "LinkCopied": "Link copied",
```

File: plugins/UsersManager/lang/en.json
Comment (sgiehl) [inline plugins/UsersManager/lang/en.json]:
```suggestion
        "InviteLinkValidDays": "Invite links are valid for %1$s days.",
```

File: plugins/CoreHome/stylesheets/coreHome.less
Comment (sgiehl) [inline plugins/CoreHome/stylesheets/coreHome.less]: Are you sure this won't have any side effects in core or for any plugin?

**Review by sgiehl — CHANGES_REQUESTED** [review]
Added a couple of comments.
And to repeat what I already mentioned on the previous PRs:
I don't think that the usability of this new feature is very good. Even I, as someone quite familiar with software having broken usability, would not expect that I first need to send an invite by mail before being able to generate a new link for copying it.
It would be so much simpler and smarter to initially provide two buttons, to either invite by mail or copying the link. That would also avoid any sense of having two separate tokens.

Also I'm not sure if only providing a button to copy the link to clipboard is the optimal solution. Some people might prefer having a textbox, where they can copy the link from.

But that's up to @justinvelluppillai and @Javi-Ormaechea, to decide if we should implement such improvements...

File: plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue
Comment (peterhashair) [reply plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue]: that's out of date, I think

File: plugins/Morpheus/stylesheets/ui/_buttons.less
Comment (peterhashair) [reply plugins/Morpheus/stylesheets/ui/_buttons.less]: that's out of date, I think

File: plugins/UsersManager/API.php
Comment (peterhashair) [reply plugins/UsersManager/API.php]: I know right, but the designer asks me to remove this.

File: plugins/CoreHome/stylesheets/coreHome.less
Comment (peterhashair) [reply plugins/CoreHome/stylesheets/coreHome.less]: I believe this is no effect from I can aware.

**Comment by peterhashair** [comment]
@justinvelluppillai any suggestion on that one? Should I go back to @Javi-Ormaechea?

File: plugins/UsersManager/API.php
Comment (sgiehl) [reply plugins/UsersManager/API.php]: Sacrificing security for usability reasons isn't anything we should consider. Did you let @Javi-Ormaechea know, that this will make Matomo bit less secure? If that's a mandatory requirement you should at least let @tsteur confirm that.

Comment (tsteur) [reply plugins/UsersManager/API.php]: From a security perspective, to invite a user via the UI we need to ask for a password confirmation for security purposes. 

I don't know where exactly we're asking for a password currently in the process. If it helps it could also look different. For example, that we only redirect to the "Invite user" screen after a password confirmation on a dedicated page. But then a user has only like 15 minutes to complete the action.

File: plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue
Comment (peterhashair) [reply plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue:526]: @sgiehl ah that was the design UI update see here. https://www.figma.com/file/ZAVoyUqPX7xUp858wOm45s/UI-Requests?node-id=57%3A2016

File: plugins/UsersManager/API.php
Comment (Javi-Ormaechea) [reply plugins/UsersManager/API.php]: @tsteur I was not aware that this was a security issue but I'm curious to know more about why a password needs to be added here when the user has technically logged in already. As an experience it makes it less seamless, and probably unexpected too (_last bit is an assumption_).

Comment (tsteur) [reply plugins/UsersManager/API.php]: @Javi-Ormaechea for sensitive actions we require a password confirmation as otherwise someone else could go onto your computer/laptop/tablet/... where you are logged in already and add a user on your behalf without your knowing. I don't know what the current screen looks like and if we ask for the password in the form. 

In some places we redirect the user for example first to a dedicated page for confirming the password and then redirect to the actual page. This password confirmation is then valid for I believe 15 minutes or so. Meaning there are options to change this.

Comment (peterhashair) [reply plugins/UsersManager/API.php]: @tsteur I can't see the password confirmation remembered for 15 minutes.

File: core/Db/Schema/Mysql.php
Comment (peterhashair) [reply core/Db/Schema/Mysql.php:58]: @sgiehl had a discussion with Justin, I guess we just go for an extra column with one expiration date for now.

File: plugins/CoreHome/stylesheets/coreHome.less
Comment (peterhashair) [reply plugins/CoreHome/stylesheets/coreHome.less]: revert that one

File: plugins/Morpheus/stylesheets/base/icons.css
Comment (peterhashair) [reply plugins/Morpheus/stylesheets/base/icons.css]: right direction added

File: plugins/UsersManager/tests/System/expected/test___UsersManager.getUsersWithSiteAccess_3_admin_when_superuseraccess.xml
Comment (peterhashair) [reply plugins/UsersManager/tests/System/expected/test___UsersManager.getUsersWithSiteAccess_3_admin_when_superuseraccess.xml]: removed token field

File: plugins/CoreHome/vue/src/Notification/Notification.ts
Comment (peterhashair) [inline plugins/CoreHome/vue/src/Notification/Notification.ts]:
```suggestion
```

**Comment by peterhashair** [comment]
Notes: I believe once that merge we need to update the docs here https://matomo.org/faq/general/manage-users/

File: plugins/UsersManager/API.php
Comment (peterhashair) [inline plugins/UsersManager/API.php]:
```suggestion
       if (Common::getRequestVar('force_api_session', 0)) {
           $this->confirmCurrentUserPassword($passwordConfirmation);
       }
```

File: plugins/Morpheus/stylesheets/base/icons.css
Comment (sgiehl) [inline plugins/Morpheus/stylesheets/base/icons.css:406]: for consistency we should maybe name the icons with a `-`.
e.g. `icon-arrow-left-2`, like you used it in demo.twig, which currently doesn't work as they have another name..

Comment (sgiehl) [inline plugins/Morpheus/stylesheets/base/icons.css]: for consistency we should maybe name the icons with a `-`.
e.g. `icon-arrow-left-2`, like you used it in demo.twig, which currently doesn't work as they have another name..

File: plugins/UsersManager/API.php
Comment (sgiehl) [inline plugins/UsersManager/API.php]: It's actually a bad solution to have a method that is able to do different things. It might not be clear for a user that the API method is either able to send a mail or returns an invite link. The method name actually only indicates the latter one. Also when changing API method, please ensure to also update the docs blocks accordingly.

File: plugins/UsersManager/API.php
Comment (sgiehl) [inline plugins/UsersManager/API.php]: An API method to return a invitation link, should for security reasons require a password when triggered through API session.

File: plugins/UsersManager/lang/en.json
Comment (sgiehl) [inline plugins/UsersManager/lang/en.json]: Changing the contents of a translations key is only suitable if it's content doesn't change too much. In that case using a new name might be better, as existing translations don't have the second placeholder.

File: plugins/UsersManager/API.php
Comment (sgiehl) [inline plugins/UsersManager/API.php]: New events should always be mentioned in the developer changelog.

File: plugins/UsersManager/vue/src/UsersManager/UsersManager.vue
Comment (sgiehl) [inline plugins/UsersManager/vue/src/UsersManager/UsersManager.vue]: I don't think it makes much sense to send a request every time the button is clicked. Some people e.g. tend to do double click on buttons. That would currently cause two simultaneous requests being sent out. In theory it could even happen that the request being resolved last, was not the one request, that updated the token last. So it could happen that you could hold a link on clipboard that is not even valid anymore.
I would suggest to disable the button, as long as a request is running. And maybe even avoid multiple requests for the same user. If the first request returned a link it should be enough to hold that link in memory every time the button is clicked - until the modal is closed.

File: core/Updates/4.12.0-b2.php
Comment (sgiehl) [inline core/Updates/4.12.0-b2.php]: Please never extend any update scripts. If you have to add changes, always increase the version in `Version.php` and use this new version for your update script. Otherwise this can cause problems if the update script was already executed for any install for any reason.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Left a couple of comments for possible improvements. Besides that I'll summarize the concerns I already mentioned in the PR(s) before:

The usability of that new "feature" really feels broken. I can't see a useful reason why you first need to send out an invitation email before being able to copy an invitation link.
Even though it might have been in the UX designs, it doesn't make sense.
Even more: To fulfill this UX requirement we now even store two different invitation tokens, so that the token sent out with the first mail is not directly invalidated when directly using the copy link feature afterwards.

That way of storing two token actually also contains a "small" security risk, as we still only have one expiry time. So sending out an email or copying a new link would magically extend the expiration of the other token. Which might be unexpected and less secure.

In terms of security we currently also only ask for a password when the user is initially invited. That makes it impossible for someone gaining access to an admin account (without knowing the password) to create a new user. But the feature to copy a new invite link currently does not ask for a password. So if there is an invited user, that has not yet accepted, an attacker would easily be able to take over that account by simply copying the invite link, without anyone noticing.


----

What I would have expected from this new feature is an implementation like:

- The form to add / invite a new user would instead of only asking for a password, show a modal similar to the resend invitation modal in this PR, but with an additional input for the password.
- This would allow to directly choose between sending an email or copying the link. Clicking one of the buttons would create the user and either send out an email or return the link to provide it to the user for copying it.
- Having the direct choice would also make the additional database field obsolete, as one token would be enough at a time.
- Clicking the copy link button should imho also show the link in a textarea or similar. There might be browsers / users where copying something to clipboard might be restricted for some reason. That would currently make it impossible to actually get the link.
- The modal to resend an invitation would actually look the same, but it would also include a password field for security reasons.

**Comment by justinvelluppillai** [comment]
@tsteur do you have any thoughts as security person on https://github.com/matomo-org/matomo/pull/19707#pullrequestreview-1107407965? We have included all security and UX/UI suggestions received so far, be good to know if there are any raised from @sgiehl that you'd like added here also.

**Comment by justinvelluppillai** [comment]
Meantime @peterhashair please proceed to address @sgiehl's other review comments, they all seem important.

**Comment by tsteur** [comment]
From what I see the password confirmation check is still there so this should be all good 👍

> That way of storing two token actually also contains a "small" security risk, as we still only have one expiry time. So sending out an email or copying a new link would magically extend the expiration of the other token. Which might be unexpected and less secure.

It's tricky. 

* We could not extend the expire time when clicking `copy link`. This could be fine as long as we show information for how long that link is valid after copying it. It's actually currently also not clear that clicking "Resent invite" will extend the 7 days. At least the wording doesn't suggest that clearly. 
* Or we still extend the time but then make it more clear that any action will extend the allowed invitation time. We just need to be clear about it at least. 

> In terms of security we currently also only ask for a password when the user is initially invited. That makes it impossible for someone gaining access to an admin account (without knowing the password) to create a new user. But the feature to copy a new invite link currently does not ask for a password. So if there is an invited user, that has not yet accepted, an attacker would easily be able to take over that account by simply copying the invite link, without anyone noticing.

It's a hard one. I would say if we're in doubt / unsure we should decide pro security. Especially since the action may be executed rarely and it's not a standard flow I would recommend asking for password confirmation to copy the link. 

fyi getting this error in the UI (not sure if it's known)

image

Unrelated: Not sure if it's important but the header still says "Add new user" when you're actually inviting a new user. It could give the indication that a user is added directly maybe. Not needed to be changed though, just an FYI that this is maybe not quite clear. Especially for existing people where they maybe don't read the buttons as they've been using them many times before.
image

I then couldn't find the feature to get the link but eventually found it by clicking on "Resend invite". I don't know what's been discussed so far and don't want to make anything more complicated. Was just thinking it could be useful if it said `Not received the invite? Resend invite or copy activation link` or something. Just a thought.

**Comment by peterhashair** [comment]
Update the PR for the following, just random thoughts unrelated. I would suggest, that we actually remove copy link, it may cause other issues later on (like copy function in js won't work). instead, we could give 2 actions which are `create a user` and `invite a user`. `create a user` goes to the old way, admin/super admin enters the password for the client. 'Invite User' only available when a client has `email_enabled` turned on.  



- Updated the resend and copy link, they both required a passwords confirmation to trigger the action, if the password was wrong, the popup will remain open until the correct password is entered

- add invite action notes, `any action will extend the invite link to x days`

**Comment by peterhashair** [comment]
As @Javi-Ormaechea requested, update the wording. It seems like our UI tester deny clipboard access, so leave the screen as that copy denied, and in order to prevent random failure with the random token, hide the token link. 

I believe there is a security risk that another site can access the clipboard which accesses the token link, the chance is rare. I think the clipboard is not meant to copy passwords etc. 

I would suggest, giving up this PR, recreating an Issue that uses `invite user` for our cloud and those who have email setup,  and `create user` as it was (admin enter password) for the user who doesn't have an email setup.

**Review by justinvelluppillai — CHANGES_REQUESTED** [review]
In safari the copied link shows up in the header when I click "Copy Link" (see screenshot):

Screen Shot 2022-09-30 at 12 23 26 PM

Also when I click Copy Link in the modal in chrome, the modal remains open but the Copy Link and Resend Invite buttons now do nothing when clicked.

**Comment by peterhashair** [comment]
@justinvelluppillai should be fixed now.

**Comment by bx80** [comment]
This is my first look at this PR, it does work :+1: and I'm aware there have already been a lot of points raised in the discussion above, I'll try to avoid rehashing previously covered comments.

a) If I want to create a new user and send them their invite link on slack, then I'd probably end up doing this:

1) Click 'Invite a new user' (goes to a screen labeled 'Add a new user')
2) Click 'Invite User'.
3) Enter my password, then 'Invitation is sent' is shown.
4) Go back to the user list.
5) Search for the newly invited user (what was their email again? Type it in...)
6) Click on an envelope row icon ("but I don't want to email, I want to copy...")
7) Click 'Copy link' on the popup.
8) Enter my password again (second time in 30 seconds :slightly_frowning_face: )
9) Finally the invite link is in the clipboard.

I did this a couple of times before I realized that the 'Resend Invite' link shown in step 3) would allow me to jump to step 7). So although I saw 'Resend Invite' link, it didn't register because I want to *copy* the link. I never wanted to *send* an invite in the first place, so wasn't intuitive. 

**Could we either rename this to 'Copy Invite Link' or duplicate the link so we have 'Resend invite' and 'Copy invite link'?** 

b) Once the 'copy link' button is used the buttons are disabled but it isn't obvious that they no longer work. If I accidentally copied something else into my clipboard and lost the copied link then clicking the 'Copy link' button again has no effect. **Wouldn't it be better to hide the buttons altogether and/or replace them with a close button?**

c) Minor cosmetic: The resend invite popup box seems to have a different font size from the normal popups? The password popup font is much smaller, it looks like part of a different application.

d) Minor point: It might be slightly clearer to say 'Link copied to the clipboard' rather than just 'Link copied'.

e) Having to enter the password twice in the space of 10 seconds isn't great, probably beyond the scope of this PR, but maybe we note this somewhere else for improvement, since we're adding password prompts to more places in the UI and risk increasing user frustration.

**Comment by peterhashair** [comment]
@bx80 I agree.
> e) Having to enter the password twice in the space of 10 seconds isn't great, probably beyond the scope of this PR, but maybe we note this somewhere else for improvement, since we're adding password prompts to more places in the UI and risk increasing user frustration.

I guess this has come to the design question. @Javi-Ormaechea 
> Minor point: It might be slightly clearer to say 'Link copied to the clipboard' rather than just 'Link copied'.

**Comment by bx80** [comment]
> >    e) Having to enter the password twice in the space of 10 seconds isn't great, probably beyond the scope of this PR, but maybe we note this somewhere else for improvement, since we're adding password prompts to more places in the UI and risk increasing user frustration.

>I guess this has come to the design question.

We could possibly look at a system level solution - same as `sudo` won't prompt for your password over and over if you entered it in the last 5 minutes. At the moment being tasked with adding 10 new users and copying their invites would make the admin type their password 20 times in under 5 minutes, it might be secure but it's the sort of UX that causes the admin to change their password to `12345` out of frustration :slightly_smiling_face:  I can log a RFC issue on this if you like?

**Comment by justinvelluppillai** [comment]
@bx80 yes it'd be good to maybe create a separate issue for that. I agree it wasn't fun getting links or sending invites during testing 👍🏽

**Review by bx80 — APPROVED** [review]
I've spent some time doing functional testing, apart from the multiple password entry annoyance which we can address elsewhere, it works as expected and allows an admin to get an invite link copied to the clipboard fairly quickly after creating the initial invite. I can't see any outstanding issues with the code.

I've pushed a couple of  UI test screenshot updates, once the test builds complete without issues then I think this is ready to merge :+1:

* [x] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [x] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [x] It's usable. [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [x] Can't see any issues [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [x] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [x] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [x] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [x] Updated [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [x] Updated  [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [x] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [x] Existing documentation updated if needed

**Comment by sgiehl** [comment]
@peterhashair There is no need to recreate pull requests when you want to target another branch. You can simply change the target branch of a PR on GitHub. Also it is important to **rebase** your changes to the new target branch.
This PR actually included some other PRs that had been merged to 4.x-dev. It doesn't hurt that those changes are now included in the release. But it's important to keep any eye on the changes your PRs include.
In general you should avoid closing PRs in order to create a new one afterwards. Unless the changes included in the PR are completely different it would be good to always use the same PR till it is merged. This ensures that all comments that were made can be directly seen. By closing and recreating a PR you actually kind of loose all comments done before, as no one will search if there might have been previous PRs.

**Comment by peterhashair** [comment]
@sgiehl yes, sorry. I confused myself there, after I recreate the PR, and release I can rebase the PR. Will avoid that in the future.

**Pull request merged**
